flink/v1.17/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/MapRangePartitioner.java

```diff
@@ -0,0 +1,288 @@
+  }
+
+  /**
+   * @return assignment summary for every subtask. Key is subtaskId. Value pair is (weight assigned
+   *     to the subtask, number of keys assigned to the subtask)
+   */
+  Map<Integer, Pair<Long, Integer>> assignmentInfo() {
+    Map<Integer, Pair<Long, Integer>> assignmentInfo = Maps.newTreeMap();
+    assignment.forEach(
+        (key, keyAssignment) -> {
+          for (int i = 0; i < keyAssignment.assignedSubtasks.length; ++i) {
+            int subtaskId = keyAssignment.assignedSubtasks[i];
+            long subtaskWeight = keyAssignment.subtaskWeights[i];
+            Pair<Long, Integer> oldValue = assignmentInfo.getOrDefault(subtaskId, Pair.of(0L, 0));
+            assignmentInfo.put(
+                subtaskId, Pair.of(oldValue.first() + subtaskWeight, oldValue.second() + 1));
+          }
+        });
+
+    return assignmentInfo;
+  }
+
+  private Map<SortKey, KeyAssignment> buildAssignment(
+      int numPartitions,
+      NavigableMap<SortKey, Long> sortedStatistics,
+      long targetWeightPerSubtask) {
+    Map<SortKey, KeyAssignment> assignmentMap =
+        Maps.newHashMapWithExpectedSize(sortedStatistics.size());
+    Iterator<SortKey> mapKeyIterator = sortedStatistics.keySet().iterator();
+    int subtaskId = 0;
+    SortKey currentKey = null;
+    long keyRemainingWeight = 0L;
+    long subtaskRemainingWeight = targetWeightPerSubtask;
+    List<Integer> assignedSubtasks = Lists.newArrayList();
+    List<Long> subtaskWeights = Lists.newArrayList();
+    while (mapKeyIterator.hasNext() && subtaskId < numPartitions) {
+      if (currentKey == null) {
+        currentKey = mapKeyIterator.next();
+        keyRemainingWeight = sortedStatistics.get(currentKey);
+      }
+
+      assignedSubtasks.add(subtaskId);
+      // assign the remaining weight of key to the current subtask if it is the last subtask
+      // or if the subtask has more capacity than the remaining key weight
+      if (subtaskId == numPartitions - 1 || keyRemainingWeight < subtaskRemainingWeight) {
+        subtaskWeights.add(keyRemainingWeight);
+        subtaskRemainingWeight -= keyRemainingWeight;
+        keyRemainingWeight = 0;
+      } else {
+        // filled up the current subtask
+        subtaskWeights.add(subtaskRemainingWeight);
+        keyRemainingWeight -= subtaskRemainingWeight;
+        // move on to the next subtask
+        subtaskId += 1;
+        subtaskRemainingWeight = targetWeightPerSubtask;
+      }
+
+      if (keyRemainingWeight <= 0) {
+        // finishing up the assignment for the current key
+        Preconditions.checkState(
+            assignedSubtasks.size() == subtaskWeights.size(),
+            "List size mismatch: assigned subtasks = %s, subtask weights = %s",
+            assignedSubtasks,
+            subtaskWeights);
+        KeyAssignment keyAssignment =
+            new KeyAssignment(
+                ArrayUtil.toIntArray(assignedSubtasks), ArrayUtil.toLongArray(subtaskWeights));
```

Comment:
I understand this is heuristic, and I am not sure the calculation worth the complexity, but here we use `weight + closeFileCost` which could be somewhat off if we have only a small amount weight left for the last subtask.
Example:
- Weight distribution for a 10 sized key, 1 closeFileCost, 10 target. The result is:
    - 9 - data + 1 closeFileCost
    - 1 - data  + 1 closeFileCost

The keyAssignment should be (9:1, but instead it will be 10:2).

Reply:
I didn't quite get the example and notations here. can you elaborate?

Reply:
Let me try again.
If we calculate with the following data:
- `closeFileCost` - 1
- `targetWeightPerSubtask` - 10
- `sortedStatsWithCloseFileCost` for the given key is - 12 (10 data + 2x1 closed file cost)

The resulting distribution based on the current calculation is `new long(10L, 2L)`
When we distribute the records, then we send 10/12 ~ 83% of the records to SORT_KEY_0, and ~16% of the records to SORT_KEY_1.

But when we are applying the distribution data, we apply it to the records, and we should not apply it to the closed file costs, so I think the correct distribution should be `new long(9L, 1L)` so we send ~ 90% to SORT_KEY_0, and 10% to SORT_KEY_1.

If we increase the `closeFileCost` compared to the fraction which gets to the last split, then the difference will be even more pronounced:
- `closeFileCost` - 90
- `targetWeightPerSubtask` - 100
- `sortedStatsWithCloseFileCost` for the given key is - 191 (11 data + 2x1 closed file cost)

The resulting distribution based on the current calculation is `new long(100L, 91L)`
When we distribute the records, then we send 100/191 ~ 52% of the records to SORT_KEY_0, and ~48% of the records to SORT_KEY_1.

I think the correct distribution should be `new long(10L, 1L)` so we send ~ 90% to SORT_KEY_0, and 10% to SORT_KEY_1.

Reply:
let me make sure I understand your example correctly

- target weight before closing cost is 9. because `SORT_KEY_0` has weight 10. it should be split into 2 trunks. hence the closing weight for the key is 2 (1x2 trunks). and total weight (including closing cost) is 12 for SORT_KEY_0.
- target weight with closing cost is 10
- SORT_KEY_0 will be assigned to 2 subtasks: subtask 0 with weight 10 and subtask 1 with weight 2

the bin packing in the last step is fine. but the `select` function should have used the weight of 9 and 1  (not 10 and 2) when route the traffic. I agree in this case, 9 and 1 are more accurate. 

I want to discuss a scenario from the unit test
```
    // target subtask weight is 10 before close file cost factored in.
    // close file cost is 2 = 20% * 10.
    // key weights before and after close file cost factored in
    // before:     35, 23, 12, 4, 1, 1, 1, 1, 1, 1
    // close-cost:  8,  6,  4, 2, 2, 2, 2, 2, 2, 2
    // after:      43, 29, 16, 6, 3, 3, 3, 3, 3, 3
    // target subtask weight per subtask is 14 (112/8)
```

for key0, the bin packing weight distribution will be 14, 14, 14, 1, because we use the total weight with closing cost. then what weights should `select` use in this case? assuming it should be the weight before closing cost. then it will be 12, 12, 12, -1/0/1? not sure about the last trunk. -1 definitely doesn't make sense. 0 also doesn't make sense.

In your example, closing cost certainly skewed the distribution away from the ideal assignment. but it might be a little more generally applicable.

Reply:
Your understanding of my example is correct.

I agree with your assessment, that it is hard to apply in some cases, but it is just highlight that these cases are calculated suboptimally.

If the `weight`-`closeFileCost` is negative, it means that we pay more in overhead for the split, than the actual data would be.

Reply:
I modified the code to ignore the residual weight if it is smaller than closing file cost. this way, the greedy algorithm doesn't need a 2nd pass